File: data_structure/binaryTree/preorder/preoder_traversal_items.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {

	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 2}}
	root.Left.Left = &TreeNode{Val: 3}
	root.Left.Right = &TreeNode{Val: 4}
	root.Right.Left = &TreeNode{Val: 4}
	root.Right.Right = &TreeNode{Val: 3}
	result := preorderTraversal(root)
	fmt.Println(result)

}

// 迭代 前序遍历二叉树
func preorderTraversal(root *TreeNode) []int {
	result := []int{}
	if root == nil {
		return result
	}
	queue := []*TreeNode{root}
	for len(queue) != 0 {
		root = queue[len(queue)- 1]
		queue = queue[:len(queue) - 1]
		if root != nil {
			result = append(result, root.Val)
			if root.Right != nil {
				queue = append(queue, root.Right)
			}
			if root.Left != nil {
				queue = append(queue, root.Left)
			}
		}
	}

	return result
}

File: data_structure/binaryTree/preorder/preoder_traversal_items.py

# 二叉树的中序遍历

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 迭代
def preorderTraversal(root: TreeNode) ->[int]:
    result = []
    if not root:
        return result
    queue = [root]
    while queue:
        root = queue.pop()
        if root:
            result.append(root.val)
            if root.right:
                queue.append(root.right)
            if root.left:
                queue.append(root.left)
    return result


if __name__ == "__main__":
    root = TreeNode(1)
    root.right = TreeNode(2)

    root.right.left = TreeNode(3)
    result = preorderTraversal(root)
    print(result)
